pytest-embedded-qemu/pytest_embedded_qemu/qemu.py

```diff
@@ -41,13 +40,16 @@ def __init__(
             raise ValueError(f'QEMU image path doesn\'t exist: {image_path}')
 
         qemu_prog_path = qemu_prog_path or self.QEMU_PROG_PATH
-        qemu_cli_args = qemu_cli_args or self.QEMU_DEFAULT_ARGS
+        qemu_cli_args = shlex.split(qemu_cli_args or self.QEMU_DEFAULT_ARGS)
+        qemu_extra_args = shlex.split(qemu_extra_args or '')
 
-        qemu_extra_args = qemu_extra_args.replace('"', '') if qemu_extra_args else qemu_extra_args
-        qemu_extra_args = [qemu_extra_args] if qemu_extra_args else []
-        qemu_extra_args.append(f'-drive file={image_path},if=mtd,format=raw')
+        super().__init__(
+            cmd=[qemu_prog_path, *qemu_cli_args, *qemu_extra_args] + ['-drive', f'file={image_path},if=mtd,format=raw'],
+            **kwargs,
+        )
 
-        cmd = f'{qemu_prog_path} {qemu_cli_args} {" ".join(qemu_extra_args)}'
-        logging.debug(cmd)
-
-        super().__init__(msg_queue, cmd, shell=True, **kwargs)
+    def _hard_reset(self):
```

Comment:
@igrr this is a drafted version. The un-implemented solution is to terminate the qemu process and re-create it with the same arguments.

Is there any better method to fake the `hard_reset` in qemu process?

Reply:
We can also send `system_reset` command to QEMU monitor (console).  This will probably require routing the console to another destination, like a TCP port: `-monitor tcp::4444,server=on,wait=off`

(more options can be found here: https://qemu-project.gitlab.io/qemu/system/invocation.html)

Reply:
And there's also QMP which is probably the better option, but I haven't personally used it yet: https://wiki.qemu.org/Documentation/QMP. There's even a Python package available to talk over QMP: https://pypi.org/project/qemu.qmp/.

Reply:
Thanks! I'll take a look.

Reply:
@igrr I added a tcp port as backend for QMP server. PTAL again!